qemu/tests/block_hotplug.py

```diff
@@ -21,144 +22,179 @@ def run(test, params, env):
     """
-    def find_disk(vm, cmd):
+    def _find_disks(session):
         """
         Find all disks in guest.
+
+        :return: The set of all disks.
+        :rtype: Set.
         """
-        if params.get("os_type") == "linux":
-            pattern = params.get("get_disk_pattern", "^/dev/vd[a-z]*$")
-        elif params.get("os_type") == "windows":
-            pattern = r"^\d+"
-            cmd = params.get("get_disk_index", "wmic diskdrive get index")
+        if os_type == "linux":
+            disks = utils_misc.list_linux_guest_disks(session)
+        elif os_type == "windows":
+            disks = set(session.cmd('wmic diskdrive get index').split()[1:])
         else:
-            test.cancel("Unsupported OS type '%s'" % params.get("os_type"))
-
-        session = vm.wait_for_login(timeout=timeout)
-        output = session.cmd_output_safe(cmd)
-        disks = re.findall(pattern, output, re.M)
-        session.close()
+            test.cancel("Unsupported OS type '%s'" % os_type)
         return disks
 
-    def get_new_disk(disk1, disk2):
+    def _get_new_disk(disk1, disk2):
         """
         Get the different disk between disk1 and disk2.
+
+        :param disk1: The set of disk1.
+        :type disk1: Set.
+        :param disk2: The set of disk2.
+        :type disk2: Set.
+        :return: A new set with disks in either the disk1 or disk2 but not both.
+        :rtype: Set
         """
-        disk = list(set(disk2).difference(set(disk1)))
-        return disk
+        return disk1 ^ disk2
 
-    def run_sub_test(params, plug_tag):
+    def run_sub_test(name):
         """
         Run subtest before/after hotplug/unplug device.
 
-        :param plug_tag: identify when to run subtest,
-                         ex, before_hotplug.
-        :return: whether vm was successfully shut-down
-                 if needed
+        :param name: identify when to run subtest, e,g, before_hotplug.
+        :return: whether vm was successfully shut-down if needed.
         """
-        sub_type = params.get("sub_type_%s" % plug_tag)
+        sub_type = params.get("sub_type_%s" % name)
         if sub_type:
-            error_context.context(context_msg % (sub_type, plug_tag),
-                                  logging.info)
+            error_context.context(
+                "Running sub test '%s' %s" % (sub_type, name), logging.info)
             utils_test.run_virt_sub_test(test, params, env, sub_type)
-            if sub_type == "shutdown" and vm.is_dead():
-                return True
-        return None
 
-    img_list = params.get("images").split()
+    def check_unplugged_disks(session, disks_before_unplug):
+        """ Check the unplungged disks. """
+        unplug_disks = utils_misc.wait_for(
+            lambda: _get_new_disk(disks_before_unplug, _find_disks(session)),
+            pause)
+        unplug_disks = set() if unplug_disks is None else unplug_disks
+        return len(unplug_disks) == extra_disks_num
```

Comment:
could use `return unplug_disk and len(unplug_disks) == extra_disks_num` to replace 71-72